# Pandas Grundlagen: DataFrames, Auswahl, Häufigkeits- und Kreuztabellen
### Unterrichtsmaterial | Julia Brutskaya

---
**Kontext:** Kurs-Umfrage der Datenwald-Akademie<br>
**Bibliotheken:** `pandas`

<img src="png%20Dateien/herr_bauchgefuehl.png" alt="Herr Bauchgefühl" width="180">


Heute Vormittag hast du die Kurs-Umfrage der Datenwald-Akademie von Hand ausgewertet: mit Strichliste, Häufigkeitstabelle und Kreuztabelle. Jetzt baust du dieselben Tabellen mit Pandas nach.

**Herr Bauchgefühl** leitet die Akademie und weiß alles, ohne nachzuschauen. Seine Behauptungen vom Vormittag tauchen hier wieder auf. Diesmal prüfst du sie mit Code.

| Block | Inhalt |
|---|---|
| A | Erste Schritte mit Pandas |
| B | Spalten und Zeilen auswählen |
| C | Filtern und Spalten bearbeiten |
| D | Datentypen und Skalenniveaus |
| E | Häufigkeitstabellen und Kreuztabellen |
| F | Weitere Übungen |

---
# Block A – Erste Schritte mit Pandas

## 1 – Was ist Pandas?

Pandas ist eine Python-Bibliothek für Daten in Tabellenform. Du kannst damit Tabellen laden, filtern, auszählen und zusammenfassen, ähnlich wie in Excel, nur mit Code.

Pandas kennt zwei Bausteine:

| Baustein | Bedeutung |
|---|---|
| **DataFrame** | eine ganze Tabelle mit Zeilen und Spalten |
| **Series** | eine einzelne Spalte |

Pandas baut auf NumPy auf, das du aus der Mathe-Woche kennst. Falls Pandas noch nicht installiert ist, führe einmalig im Terminal aus:

```bash
pip install pandas
```

In [70]:
# pandas wird fast immer unter dem Kurznamen "pd" importiert
import pandas as pd

print("Pandas-Version:", pd.__version__)

Pandas-Version: 3.0.6


## 2 – Die Kurs-Umfrage als DataFrame

Die Umfrage vom Vormittag steht hier als Liste. Jeder Eintrag in runden Klammern ist eine befragte Person. `pd.DataFrame()` macht daraus eine Tabelle, die Liste `spalten` liefert die Spaltennamen.

In [71]:
# Kurs-Umfrage der Datenwald-Akademie: eine Zeile pro befragter Person
zeilen = [
    (101, "Vormittag", "Kaffee", "M", 0, 25), (102, "Abend", "Tee", "S", 1, 35),
    (103, "Vormittag", "Mate", "L", 0, 18), (104, "Vormittag", "Kaffee", "L", 2, 42),
    (105, "Abend", "Wasser", "M", 0, 30), (106, "Vormittag", "Tee", "M", 1, 28),
    (107, "Abend", "Kaffee", "XL", 0, 55), (108, "Vormittag", "Kaffee", "S", 3, 12),
    (109, "Vormittag", "Mate", "M", 1, 38), (110, "Abend", "Tee", "M", 0, 115),
    (111, "Vormittag", "Kaffee", "L", 0, 32), (112, "Abend", "Mate", "L", 2, 45),
    (113, "Vormittag", "Wasser", "S", 1, 22), (114, "Vormittag", "Kaffee", "M", 0, 35),
    (115, "Abend", "Tee", "L", 1, 48), (116, "Vormittag", "Tee", "M", 0, 30),
    (117, "Abend", "Kaffee", "M", 3, 65), (118, "Vormittag", "Mate", "XL", 1, 40),
    (119, "Abend", "Wasser", "S", 0, 52), (120, "Vormittag", "Kaffee", "L", 2, 25),
]
spalten = ["teilnehmer_id", "kurszeit", "getraenk", "tshirt", "haustiere", "anfahrt_min"]

umfrage = pd.DataFrame(zeilen, columns=spalten)
umfrage

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
0,101,Vormittag,Kaffee,M,0,25
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18
3,104,Vormittag,Kaffee,L,2,42
4,105,Abend,Wasser,M,0,30
5,106,Vormittag,Tee,M,1,28
6,107,Abend,Kaffee,XL,0,55
7,108,Vormittag,Kaffee,S,3,12
8,109,Vormittag,Mate,M,1,38
9,110,Abend,Tee,M,0,115


Die Begriffe vom Vormittag findest du in der Tabelle wieder:

| Statistik | Pandas |
|---|---|
| Merkmalsträger | eine Zeile |
| Merkmal (Variable) | eine Spalte |
| Merkmalsausprägung | der Wert in einer Zelle |
| Urliste | alle Werte einer Spalte in der erhobenen Reihenfolge |

Die Zahlen ganz links (0 bis 19) gehören nicht zu den Daten. Das ist der **Index**, mit dem Pandas die Zeilen durchnummeriert.

## 3 – Den DataFrame inspizieren

Bevor du mit einer Tabelle rechnest, verschaffst du dir einen Überblick.

| Befehl | Zeigt |
|---|---|
| `umfrage.head(n)` | die ersten n Zeilen (ohne Angabe: 5) |
| `umfrage.tail(n)` | die letzten n Zeilen |
| `umfrage.shape` | Anzahl der Zeilen und Spalten |
| `umfrage.columns` | die Spaltennamen |
| `umfrage.info()` | Spalten, Datentypen und fehlende Werte |

In [72]:
# Die ersten 3 Zeilen
umfrage.head(3)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
0,101,Vormittag,Kaffee,M,0,25
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18


In [ ]:
# Die letzten 2 Zeilen
umfrage.tail(2)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
15,116,Vormittag,Tee,M,0,30
16,117,Abend,Kaffee,M,3,65
17,118,Vormittag,Mate,XL,1,40
18,119,Abend,Wasser,S,0,52
19,120,Vormittag,Kaffee,L,2,25


In [177]:
# shape liefert ein Tupel: (Zeilen, Spalten)
print(umfrage.shape)

anzahl_zeilen, anzahl_spalten = umfrage.shape
print(f"{anzahl_zeilen} Merkmalsträger, {anzahl_spalten} Merkmale")

(20, 6)
20 Merkmalsträger, 6 Merkmale


`shape` ist ein Attribut und keine Methode. Deshalb stehen dahinter keine Klammern.

In [75]:
# Spaltennamen als Python-Liste
print(umfrage.columns.tolist())

['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_min']


In [76]:
# Technische Zusammenfassung: Spalten, Anzahl vorhandener Werte, Datentypen
umfrage.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   teilnehmer_id  20 non-null     int64
 1   kurszeit       20 non-null     str  
 2   getraenk       20 non-null     str  
 3   tshirt         20 non-null     str  
 4   haustiere      20 non-null     int64
 5   anfahrt_min    20 non-null     int64
dtypes: int64(3), str(3)
memory usage: 1.1 KB


`Non-Null Count` zeigt, wie viele Werte in einer Spalte vorhanden sind. Hier sind es überall 20, es fehlt also nichts.

### Übung 1

- **a)** Zeige die ersten 7 Zeilen der Umfrage.
- **b)** Zeige die letzten 4 Zeilen.
- **c)** Gib mit einem f-String den Satz „Die Umfrage hat ... Zeilen." aus. Die Zahl soll aus `shape` kommen.
- **d)** Wie viele Spalten enthalten Zahlen, wie viele Text? Lies es aus `info()` ab und schreibe die Antwort als Kommentar.

In [178]:
# a)
umfrage.head(7)


,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
0,101,Vormittag,Kaffee,M,0,25
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18
3,104,Vormittag,Kaffee,L,2,42
4,105,Abend,Wasser,M,0,30
5,106,Vormittag,Tee,M,1,28
6,107,Abend,Kaffee,XL,0,55


In [179]:
# b)
umfrage.tail(4)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
16,117,Abend,Kaffee,M,3,65
17,118,Vormittag,Mate,XL,1,40
18,119,Abend,Wasser,S,0,52
19,120,Vormittag,Kaffee,L,2,25


In [181]:
# c)
print(f"Die Umfrage hat {umfrage.shape[0]} Zeilen.")

Die Umfrage hat 20 Zeilen.


In [182]:
# d)
umfrage.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   teilnehmer_id  20 non-null     int64   
 1   kurszeit       20 non-null     str     
 2   getraenk       20 non-null     str     
 3   tshirt         20 non-null     category
 4   haustiere      20 non-null     int64   
 5   anfahrt_min    20 non-null     int64   
dtypes: category(1), int64(3), str(2)
memory usage: 1.1 KB


---
# Block B – Spalten und Zeilen auswählen

<img src="png%20Dateien/nils_eichhoernchen.png" alt="Nils, das Eichhörnchen" width="170">

**Nils, das Eichhörnchen,** zählt und misst. Ihn interessieren vor allem die Zahlenspalten. Dafür muss er sie erst einmal aus der Tabelle herausholen.

## 4 – Eine Spalte auswählen

Mit eckigen Klammern und dem Spaltennamen holst du eine Spalte aus dem DataFrame. Das Ergebnis ist eine **Series**.

In [81]:
getraenke = umfrage["getraenk"]

print(type(getraenke))
getraenke.head()

<class 'pandas.Series'>


0    Kaffee
1       Tee
2      Mate
3    Kaffee
4    Wasser
Name: getraenk, dtype: str

Mit `.tolist()` wird aus der Series eine Python-Liste. Das ist die **Urliste** vom Vormittag.

In [82]:
print(umfrage["getraenk"].tolist())

['Kaffee', 'Tee', 'Mate', 'Kaffee', 'Wasser', 'Tee', 'Kaffee', 'Kaffee', 'Mate', 'Tee', 'Kaffee', 'Mate', 'Wasser', 'Kaffee', 'Tee', 'Tee', 'Kaffee', 'Mate', 'Wasser', 'Kaffee']


Für mehrere Spalten übergibst du eine Liste von Namen. Achte auf die doppelten eckigen Klammern. Das Ergebnis ist wieder ein DataFrame.

In [184]:
umfrage[["teilnehmer_id", "getraenk", "anfahrt_min"]].head()

,teilnehmer_id,getraenk,anfahrt_min
0,101,Kaffee,25
1,102,Tee,35
2,103,Mate,18
3,104,Kaffee,42
4,105,Wasser,30


## 5 – Kennzahlen einer Spalte

> **Herr Bauchgefühl:** „Alle brauchen ungefähr eine halbe Stunde zur Akademie."

Nils misst nach. Jede Series bringt Methoden mit, die alle Werte der Spalte zu einer Zahl zusammenfassen.

In [84]:
print("Kürzeste Anfahrt:", umfrage["anfahrt_min"].min(), "Minuten")
print("Längste Anfahrt: ", umfrage["anfahrt_min"].max(), "Minuten")

Kürzeste Anfahrt: 12 Minuten
Längste Anfahrt:  115 Minuten


Die Anfahrt reicht von 12 bis 115 Minuten. Von „ungefähr einer halben Stunde" kann keine Rede sein. Herr Bauchgefühl stutzt.

<img src="png%20Dateien/der_ausreisser.png" alt="Der Ausreißer, der Kater" width="160">

Die 115 Minuten liegen weit weg vom Rest. So einen Wert nennt man **Ausreißer**. Der Kater ist schon wieder verschwunden, er kommt später in der Woche zurück.

| Methode | Ergebnis |
|---|---|
| `.min()` | kleinster Wert |
| `.max()` | größter Wert |
| `.sum()` | Summe aller Werte |
| `.count()` | Anzahl der vorhandenen Werte |

In [85]:
print("Haustiere insgesamt:", umfrage["haustiere"].sum())
print("Befragte Personen:  ", umfrage["haustiere"].count())

Haustiere insgesamt: 18
Befragte Personen:   20


## 6 – Zeilen auswählen mit `.iloc` und `.loc`

| Methode | Auswahl über | Merkhilfe |
|---|---|---|
| `.iloc[]` | die **Position** (0, 1, 2, ...) | *integer location* |
| `.loc[]` | das **Label** im Index und den Spaltennamen | *label* |

In [86]:
# Die Zeile an Position 0, also die erste Zeile
umfrage.iloc[0]

teilnehmer_id          101
kurszeit         Vormittag
getraenk            Kaffee
tshirt                   M
haustiere                0
anfahrt_min             25
Name: 0, dtype: object

In [87]:
# Positionen 0 bis 2: Das Ende (3) ist ausgeschlossen, wie bei Listen
umfrage.iloc[0:3]

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
0,101,Vormittag,Kaffee,M,0,25
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18


In [88]:
# Ein einzelner Wert: Zeile an Position 0, Spalte an Position 2
print(umfrage.iloc[0, 2])

Kaffee


In [89]:
# Labels 0 bis 3 und zwei Spalten nach Namen: Das Ende (3) ist hier eingeschlossen
umfrage.loc[0:3, ["kurszeit", "getraenk"]]

,kurszeit,getraenk
0,Vormittag,Kaffee
1,Abend,Tee
2,Vormittag,Mate
3,Vormittag,Kaffee


Auch `.loc` holt einen einzelnen Wert. Und `.iloc` nimmt Bereiche für Zeilen und Spalten gleichzeitig.

In [90]:
# Ein einzelner Wert über Label und Spaltenname
print(umfrage.loc[0, "getraenk"])

Kaffee


In [91]:
# Zeilen an Position 1 bis 3, Spalten an Position 0 und 1
umfrage.iloc[1:4, 0:2]

,teilnehmer_id,kurszeit
1,102,Abend
2,103,Vormittag
3,104,Vormittag



**Achtung:** Bei `.iloc` ist das Ende eines Bereichs ausgeschlossen, bei `.loc` ist es eingeschlossen.

Solange der Index einfach 0, 1, 2, ... lautet, bezeichnen Position und Label dieselbe Zeile. In Abschnitt 11 ändert sich das.

### Übung 2

- **a)** Wie viele Haustiere hat die Person mit den meisten Haustieren?
- **b)** Zeige für die ersten fünf Personen nur die Spalten `teilnehmer_id` und `anfahrt_min`.
- **c)** Hole mit `.iloc` den Wert in der Zeile an Position 9 und der Spalte an Position 5. Welche Anfahrt ist das?

In [185]:
# a)
print(umfrage["haustiere"].max())

3


In [190]:
# b)
# umfrage[["teilnehmer_id", "anfahrt_min"]].head()
umfrage.loc[0:4, ["teilnehmer_id", "anfahrt_min"]]

,teilnehmer_id,anfahrt_min
0,101,25
1,102,35
2,103,18
3,104,42
4,105,30


In [ ]:
# Ausreißer - anfahrt_min
print(umfrage.iloc[9,5])

115


---
# Block C – Filtern und Spalten bearbeiten

## 7 – Zeilen filtern mit Bedingungen

Eine Bedingung auf einer Spalte ergibt eine Series aus `True` und `False`. Man nennt sie **Maske**.

In [95]:
maske = umfrage["anfahrt_min"] > 45
print(maske.head(8))

0    False
1    False
2    False
3    False
4    False
5    False
6     True
7    False
Name: anfahrt_min, dtype: bool


Setzt du die Maske in eckige Klammern, bleiben nur die Zeilen mit `True` übrig.

In [96]:
umfrage[maske]

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
6,107,Abend,Kaffee,XL,0,55
9,110,Abend,Tee,M,0,115
14,115,Abend,Tee,L,1,48
16,117,Abend,Kaffee,M,3,65
18,119,Abend,Wasser,S,0,52


Mehrere Bedingungen verbindest du mit `&` (und) oder `|` (oder). Jede Bedingung braucht eigene runde Klammern.

In [97]:
# Vormittagskurs UND mindestens zwei Haustiere
umfrage[(umfrage["kurszeit"] == "Vormittag") & (umfrage["haustiere"] >= 2)]

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
3,104,Vormittag,Kaffee,L,2,42
7,108,Vormittag,Kaffee,S,3,12
19,120,Vormittag,Kaffee,L,2,25


`.isin([...])` prüft, ob ein Wert in einer Liste vorkommt. Das spart mehrere Bedingungen mit `|`.

In [98]:
# Alle, die Tee oder Mate trinken
umfrage[umfrage["getraenk"].isin(["Tee", "Mate"])].head()

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18
5,106,Vormittag,Tee,M,1,28
8,109,Vormittag,Mate,M,1,38
9,110,Abend,Tee,M,0,115


## 8 – Anteile berechnen

<img src="png%20Dateien/herr_bauchgefuehl.png" alt="Herr Bauchgefühl" width="150">

> **Herr Bauchgefühl:** „Die drei an der Kaffeemaschine lieben Kaffee. Also lieben alle Kaffee."

Er hat die Personen mit den IDs 101, 104 und 108 gefragt. Das ist seine Stichprobe.

In [99]:
kaffeemaschine = umfrage[umfrage["teilnehmer_id"].isin([101, 104, 108])]
kaffeemaschine[["teilnehmer_id", "getraenk"]]

,teilnehmer_id,getraenk
0,101,Kaffee
3,104,Kaffee
7,108,Kaffee


Dreimal Kaffee, in dieser Stichprobe also 100 %. Wer an der Kaffeemaschine fragt, trifft aber vor allem Kaffeetrinker. Das ist eine bequeme Auswahl und damit verzerrt.

Jetzt dieselbe Frage für alle 20 Befragten:

In [100]:
trinkt_kaffee = umfrage["getraenk"] == "Kaffee"

print("Anzahl:", trinkt_kaffee.sum())
print("Anteil:", trinkt_kaffee.mean())
print(f"In Prozent: {trinkt_kaffee.mean():.0%}")

Anzahl: 8
Anteil: 0.4
In Prozent: 40%


Warum funktioniert das? Pandas rechnet `True` als 1 und `False` als 0.

- `.sum()` zählt deshalb, wie oft die Bedingung zutrifft.
- `.mean()` liefert den Anteil, bei dem sie zutrifft.

Die Angabe `:.0%` im f-String formatiert die Zahl als Prozentwert ohne Nachkommastellen.

## 9 – Zufallsstichproben ziehen

`.sample(n=5)` zieht fünf zufällige Zeilen. Der Parameter `random_state` legt den Startwert des Zufallsgenerators fest, damit das Ergebnis wiederholbar ist.

In [101]:
for startwert in [1, 2, 6]:
    stichprobe = umfrage.sample(n=5, random_state=startwert)
    anteil = (stichprobe["getraenk"] == "Kaffee").mean()
    print(f"Startwert {startwert}: {anteil:.0%} Kaffee")

Startwert 1: 80% Kaffee
Startwert 2: 20% Kaffee
Startwert 6: 40% Kaffee


Jede Stichprobe liefert einen anderen Anteil, obwohl unter allen 20 Befragten immer 40 % Kaffee trinken. Eine Zufallsstichprobe ist fair, aber fünf Personen sind für eine verlässliche Aussage zu wenig.

## 10 – Spalten hinzufügen, umbenennen und löschen

Eine neue Spalte entsteht durch eine Zuweisung. Pandas rechnet dabei Zeile für Zeile.

In [192]:
# Anfahrt in Stunden statt in Minuten
umfrage["anfahrt_std"] = round(umfrage["anfahrt_min"] / 60, 2)
umfrage.head(3)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min,anfahrt_std
0,101,Vormittag,Kaffee,M,0,25,0.42
1,102,Abend,Tee,S,1,35,0.58
2,103,Vormittag,Mate,L,0,18,0.30


`.rename()` und `.drop()` geben einen **neuen** DataFrame zurück. Das Original ändert sich erst, wenn du das Ergebnis wieder zuweist.

In [103]:
# rename erwartet ein Dictionary: {alter_name: neuer_name}
umbenannt = umfrage.rename(columns={"anfahrt_std": "anfahrt_stunden"})

print("umbenannt:", umbenannt.columns.tolist())
print("umfrage:  ", umfrage.columns.tolist())

umbenannt: ['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_min', 'anfahrt_stunden']
umfrage:   ['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_min', 'anfahrt_std']


In [104]:
# Die Hilfsspalte wieder entfernen und das Ergebnis zuweisen
umfrage = umfrage.drop(columns=["anfahrt_std"])
print(umfrage.columns.tolist())

['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_min']


**Hinweis:** In fremdem Code siehst du oft `inplace=True`, zum Beispiel `umfrage.drop(columns=["anfahrt_std"], inplace=True)`. Das ändert den DataFrame direkt, ohne Zuweisung.

## 11 – Den Index setzen

Statt 0, 1, 2, ... kann auch eine Spalte als Index dienen. Dann sprichst du Zeilen über ihre Werte an.

In [105]:
nach_id = umfrage.set_index("teilnehmer_id")
nach_id.head(3)

,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
teilnehmer_id,,,,,
101,Vormittag,Kaffee,M,0,25
102,Abend,Tee,S,1,35
103,Vormittag,Mate,L,0,18


Jetzt unterscheiden sich `.loc` und `.iloc`: `.loc[110]` sucht das **Label** 110, `.iloc[9]` die **Position** 9. Beides ist dieselbe Person.

In [106]:
print(nach_id.loc[110])
print()
print(nach_id.iloc[9])

kurszeit       Abend
getraenk         Tee
tshirt             M
haustiere          0
anfahrt_min      115
Name: 110, dtype: object

kurszeit       Abend
getraenk         Tee
tshirt             M
haustiere          0
anfahrt_min      115
Name: 110, dtype: object


In [107]:
# reset_index macht aus dem Index wieder eine normale Spalte
nach_id.reset_index().head(3)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min
0,101,Vormittag,Kaffee,M,0,25
1,102,Abend,Tee,S,1,35
2,103,Vormittag,Mate,L,0,18


### Übung 3

- **a)** Zeige alle Personen aus dem Abendkurs, die länger als 45 Minuten unterwegs sind.
- **b)** Welcher Anteil der Befragten hat mindestens ein Haustier? Gib ihn in Prozent aus.
- **c)** Erstelle einen DataFrame `umbenannt`, in dem die Spalte `anfahrt_min` den Namen `anfahrt_minuten` trägt. `umfrage` selbst soll unverändert bleiben. Gib die Spaltennamen beider DataFrames aus.

In [193]:
# a)
umfrage[(umfrage["kurszeit"] == "Abend") & (umfrage["anfahrt_min"] > 45)]

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min,anfahrt_std
6,107,Abend,Kaffee,XL,0,55,0.92
9,110,Abend,Tee,M,0,115,1.92
14,115,Abend,Tee,L,1,48,0.80
16,117,Abend,Kaffee,M,3,65,1.08
18,119,Abend,Wasser,S,0,52,0.87


In [195]:
# b)
hat_tier = umfrage["haustiere"] > 0
print(f"Mindestens ein Haustier: {hat_tier.mean():.0%}")

Mindestens ein Haustier: 55%


In [197]:
# c)"
umbenannt = umfrage.rename(columns={"anfahrt_min": "anfahrt_minuten"})
print(umbenannt.columns.tolist())
print(umfrage.columns.tolist())

['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_minuten', 'anfahrt_std']
['teilnehmer_id', 'kurszeit', 'getraenk', 'tshirt', 'haustiere', 'anfahrt_min', 'anfahrt_std']


---
# Block D – Datentypen und Skalenniveaus

<img src="png%20Dateien/etta_elster.png" alt="Etta, die Elster" width="170">

**Etta, die Elster,** sortiert und etikettiert. Sie achtet darauf, welche Spalten nur Etiketten enthalten und mit welchen man wirklich rechnen darf.

## 12 – Datentyp ist nicht Skalenniveau

> **Herr Bauchgefühl:** „Ich hätte gern die durchschnittliche Teilnehmer-ID."

In [111]:
print(umfrage["teilnehmer_id"].mean())

110.5


Pandas rechnet 110.5 aus, ohne zu protestieren. Inhaltlich ist die Zahl sinnlos: Die ID ist ein Etikett aus Ziffern, also nominal.

In [112]:
print(umfrage.dtypes)

teilnehmer_id    int64
kurszeit           str
getraenk           str
tshirt             str
haustiere        int64
anfahrt_min      int64
dtype: object


Pandas kennt nur den **Datentyp** einer Spalte. Das **Skalenniveau** musst du selbst wissen.

| Spalte | Datentyp in Pandas | Skalenniveau |
|---|---|---|
| `teilnehmer_id` | Ganzzahl (`int64`) | nominal |
| `kurszeit` | Text | nominal |
| `getraenk` | Text | nominal |
| `tshirt` | Text | ordinal |
| `haustiere` | Ganzzahl (`int64`) | Verhältnis, diskret |
| `anfahrt_min` | Ganzzahl (`int64`) | Verhältnis, stetig |

Textspalten zeigt Pandas je nach Version als `str` oder als `object` an.

## 13 – Ausprägungen ansehen

`.unique()` listet jede Ausprägung einmal auf, `.nunique()` zählt sie. Mit `.tolist()` wird die Ausgabe zu einer einfachen Liste.

In [113]:
print(umfrage["getraenk"].unique().tolist())
print("Anzahl verschiedener Getränke:", umfrage["getraenk"].nunique())

['Kaffee', 'Tee', 'Mate', 'Wasser']
Anzahl verschiedener Getränke: 4


Die T-Shirt-Größen sind ordinal: S < M < L < XL. Pandas weiß davon nichts und sortiert Text alphabetisch.

In [198]:
print(sorted(umfrage["tshirt"].unique()))

['L', 'M', 'S', 'XL']


## 14 – Geordnete Kategorien mit `pd.Categorical`

Etta klebt die Etiketten in der richtigen Reihenfolge auf. Mit `pd.Categorical` teilst du Pandas die Reihenfolge mit: `categories` nennt die Ausprägungen von klein nach groß, `ordered=True` schaltet den Vergleich ein.

In [210]:
groessen = ["S", "M", "L", "XL"]
umfrage["tshirt"] = pd.Categorical(umfrage["tshirt"], categories=groessen, ordered=True)

print(umfrage["tshirt"].dtype)

category


In [201]:
print("Kleinste Größe:", umfrage["tshirt"].min())
print("Größte Größe:  ", umfrage["tshirt"].max())
print("Größe M oder kleiner:", (umfrage["tshirt"] <= "M").sum(), "Personen")

Kleinste Größe: S
Größte Größe:   XL
Größe M oder kleiner: 12 Personen


Auch das Sortieren richtet sich jetzt nach der Größe. `sort_values("spalte")` sortiert aufsteigend, mit `ascending=False` absteigend.

In [206]:
umfrage.sort_values("tshirt").head(6)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min,anfahrt_std
1,102,Abend,Tee,S,1,35,0.58
12,113,Vormittag,Wasser,S,1,22,0.37
18,119,Abend,Wasser,S,0,52,0.87
7,108,Vormittag,Kaffee,S,3,12,0.20
0,101,Vormittag,Kaffee,M,0,25,0.42
16,117,Abend,Kaffee,M,3,65,1.08


### Übung 4

- **a)** Berechne den Mittelwert von `haustiere` und den Mittelwert von `teilnehmer_id`. Schreibe als Kommentar dazu, welcher der beiden Werte inhaltlich sinnvoll ist und warum.
- **b)** Wie viele Personen tragen Größe L oder größer?
- **c)** Sortiere die Umfrage absteigend nach `tshirt` und zeige die ersten drei Zeilen.

In [ ]:
# a)
print("Haustiere:", umfrage["haustiere"].mean())        # Verhältnisskala
print("Teilnehmer-ID:", umfrage["teilnehmer_id"].mean())

Haustiere: 0.9
Teilnehmer-ID: 110.5


In [212]:
# b)
print("Größe L oder größer:", (umfrage["tshirt"] >= "L").sum())

Größe L oder größer: 8


In [213]:
# c)
umfrage.sort_values("tshirt", ascending=False).head(3)

,teilnehmer_id,kurszeit,getraenk,tshirt,haustiere,anfahrt_min,anfahrt_std
6,107,Abend,Kaffee,XL,0,55,0.92
17,118,Vormittag,Mate,XL,1,40,0.67
10,111,Vormittag,Kaffee,L,0,32,0.53


---
# Block E – Häufigkeitstabellen und Kreuztabellen

<img src="png%20Dateien/nils_eichhoernchen.png" alt="Nils, das Eichhörnchen" width="170">

Vormittags hat Nils Strichlisten geführt. Pandas zählt schneller, und das Ergebnis kennst du schon.

## 15 – Absolute und relative Häufigkeit

> **Herr Bauchgefühl:** „Acht Leute trinken Kaffee. Das ist die Mehrheit!"

`.value_counts()` zählt, wie oft jede Ausprägung vorkommt. Das ist die **absolute Häufigkeit**.

In [121]:
print(umfrage["getraenk"].value_counts())

getraenk
Kaffee    8
Tee       5
Mate      4
Wasser    3
Name: count, dtype: int64


Mit `normalize=True` teilt Pandas jede Anzahl durch die Gesamtzahl. Das ist die **relative Häufigkeit**.

In [122]:
print(umfrage["getraenk"].value_counts(normalize=True))

getraenk
Kaffee    0.40
Tee       0.25
Mate      0.20
Wasser    0.15
Name: proportion, dtype: float64


Kaffee ist die größte Gruppe, mit 40 % aber nicht die Mehrheit: 60 % trinken etwas anderes. Das zeigt erst die relative Häufigkeit. Herr Bauchgefühl stutzt.

`.value_counts()` sortiert nach Häufigkeit, die häufigste Ausprägung steht oben.

## 16 – Eine vollständige Häufigkeitstabelle

Für ordinale Merkmale kommt die **kumulierte Häufigkeit** dazu.

- `.sort_index()` sortiert nach der Ausprägung (S, M, L, XL) statt nach der Häufigkeit.
- `.cumsum()` zählt Zeile für Zeile auf.

In [123]:
absolut = umfrage["tshirt"].value_counts().sort_index()

tabelle = pd.DataFrame({"absolut": absolut})
tabelle["relativ"] = tabelle["absolut"] / tabelle["absolut"].sum()
tabelle["kumuliert"] = tabelle["absolut"].cumsum()
tabelle["kumuliert_relativ"] = tabelle["relativ"].cumsum()
tabelle

,absolut,relativ,kumuliert,kumuliert_relativ
tshirt,,,,
S,4,0.2,4,0.2
M,8,0.4,12,0.6
L,6,0.3,18,0.9
XL,2,0.1,20,1.0


Lesebeispiel: 12 von 20 Personen, also 60 %, tragen Größe M oder kleiner.

Kumulieren ist erst ab ordinalem Skalenniveau sinnvoll. Bei den Getränken gibt es keine Reihenfolge, auf die sich „bis einschließlich" beziehen könnte.

## 17 – Klassen bilden mit `pd.cut`

Ein stetiges Merkmal wie die Anfahrt hat fast so viele Ausprägungen wie Personen. `pd.cut` teilt die Werte in Klassen ein, `bins` nennt die Klassengrenzen.

Die Schreibweise kennst du vom Vormittag: `(20, 40]` bedeutet über 20 bis einschließlich 40.

In [124]:
klassen = pd.cut(umfrage["anfahrt_min"], bins=[0, 20, 40, 60, 120])
print(klassen.value_counts().sort_index())

anfahrt_min
(0, 20]       2
(20, 40]     11
(40, 60]      5
(60, 120]     2
Name: count, dtype: int64


## 18 – Kreuztabellen mit `pd.crosstab`

> **Herr Bauchgefühl:** „Kaffee ist überall die Nummer eins, ob Vormittags- oder Abendkurs."

Eine Kreuztabelle zählt zwei Merkmale gleichzeitig aus. Das erste Merkmal bildet die Zeilen, das zweite die Spalten. `margins=True` ergänzt die Randhäufigkeiten.

In [125]:
pd.crosstab(umfrage["kurszeit"], umfrage["getraenk"], margins=True, margins_name="Gesamt")

getraenk,Kaffee,Mate,Tee,Wasser,Gesamt
kurszeit,,,,,
Abend,2,1,3,2,8
Vormittag,6,3,2,1,12
Gesamt,8,4,5,3,20


Im Abendkurs trinken 3 Personen Tee und nur 2 Kaffee. Weil die Kurse verschieden groß sind (8 und 12 Personen), vergleichst du besser **Zeilenprozente**: Jede Zeile ergibt zusammen 100 %.

In [126]:
zeilenprozente = pd.crosstab(umfrage["kurszeit"], umfrage["getraenk"], normalize="index") * 100
zeilenprozente.round(1)

getraenk,Kaffee,Mate,Tee,Wasser
kurszeit,,,,
Abend,25.0,12.5,37.5,25.0
Vormittag,50.0,25.0,16.7,8.3


Im Abendkurs führt Tee mit 37,5 %, Kaffee kommt nur auf 25 %. Herr Bauchgefühl stutzt ein letztes Mal.

| Vom Vormittag | In Pandas |
|---|---|
| Zeilenprozente | `normalize="index"` |
| Spaltenprozente | `normalize="columns"` |
| Gesamtprozente | `normalize="all"` |

### Übung 5

- **a)** Erstelle für `kurszeit` die absolute und die relative Häufigkeit.
- **b)** Erstelle eine Kreuztabelle aus `kurszeit` und `tshirt` mit Randhäufigkeiten.
- **c)** Wie viel Prozent der Personen mit Größe S sind im Abendkurs? Nutze Spaltenprozente.

In [216]:
# a)
print(umfrage["kurszeit"].value_counts())
print()
print(umfrage["kurszeit"].value_counts(normalize=True))

kurszeit
Vormittag    12
Abend         8
Name: count, dtype: int64

kurszeit
Vormittag    0.6
Abend        0.4
Name: proportion, dtype: float64


In [217]:
# b)
pd.crosstab(umfrage["kurszeit"], umfrage["tshirt"], margins=True, margins_name="Gesamt")

tshirt,S,M,L,XL,Gesamt
kurszeit,,,,,
Abend,2,3,2,1,8
Vormittag,2,5,4,1,12
Gesamt,4,8,6,2,20


In [218]:
# c)
spaltenprozente = pd.crosstab(umfrage["kurszeit"], umfrage["tshirt"], normalize="columns")
spaltenprozente.round(1)


tshirt,S,M,L,XL
kurszeit,,,,
Abend,0.5,0.4,0.3,0.5
Vormittag,0.5,0.6,0.7,0.5


---
# Block F – Weitere Übungen

Jede Übung nennt den Block, zu dem sie gehört.

- **Übung 6 bis 9:** Du prüfst deine Handergebnisse vom Vormittag mit Pandas nach.
- **Übung 10:** Ein neuer Datensatz, die Charts.
- **Übung 11 bis 15:** Ein zweiter Schauplatz, die TechAcademy GmbH.
- **Übung 16:** Zusatz, Series und DataFrames selbst bauen.

In [174]:
# sep="\t": Die Spalten sind in dieser Datei durch Tabulatoren getrennt
charts = pd.read_csv("csv & excel Dateien/UK-top40-1964-1-2.tsv", sep="\t")
charts.head()

,Pos,LW,Title,Artist,Publisher,Peak Pos,WoC
0,1,1,I WANT TO HOLD YOUR HAND,THE BEATLES,PARLOPHONE,1,5
1,2,6,GLAD ALL OVER,THE DAVE CLARK FIVE,COLUMBIA,2,7
2,3,2,SHE LOVES YOU,THE BEATLES,PARLOPHONE,1,19
3,4,3,YOU WERE MADE FOR ME,FREDDIE AND THE DREAMERS,COLUMBIA,3,9
4,5,9,TWENTY FOUR HOURS FROM TULSA,GENE PITNEY,UNITED ARTISTS,5,5


## 19 – Die Handübungen vom Vormittag nachprüfen

Arbeite weiter mit dem DataFrame `umfrage`. Lege deine Lösungen vom Vormittag daneben und vergleiche.

### Übung 6 – Haustiere auszählen (zu Block E, Häufigkeitstabellen)

Vormittags war das Übung 7. Erstelle für `haustiere` eine Häufigkeitstabelle mit absoluter, relativer und kumulierter relativer Häufigkeit. Stimmt sie mit deiner Strichliste überein?

In [130]:
# Übung 6


### Übung 7 – Eigene Klassen (zu Block E, Häufigkeitstabellen)

Vormittags war das Übung 8. Teile `anfahrt_min` in die Klassen (0, 30], (30, 60] und (60, 120] ein und zähle aus, wie viele Personen in jede Klasse fallen.

In [131]:
# Übung 7


### Übung 8 – Haustier und Kurszeit (zu Block C und E, Kreuztabellen)

Vormittags war das Übung 9.

- **a)** Füge der Umfrage eine Spalte `hat_haustier` hinzu. Sie soll `True` enthalten, wenn eine Person mindestens ein Haustier hat.
- **b)** Erstelle eine Kreuztabelle aus `kurszeit` und `hat_haustier` mit Randhäufigkeiten.
- **c)** Wie viel Prozent aller Befragten sind im Abendkurs und haben ein Haustier? Nutze Gesamtprozente.
- **d)** In welchem Kurs ist der Anteil der Personen mit Haustier größer? Nutze Zeilenprozente.

In [132]:
# a)


In [133]:
# b)


In [134]:
# c)


In [135]:
# d)


### Übung 9 – Die Mate-Fans (zu Block E, Kreuztabellen)

Vormittags war das Übung 10.

- **a)** Wie viel Prozent der Mate-Fans sind im Vormittagskurs?
- **b)** Wie viel Prozent des Vormittagskurses trinken Mate?

Schreibe als Kommentar dazu, welche Prozentart du jeweils gebraucht hast.

In [136]:
# a)


In [137]:
# b)


## 20 – Ein neuer Datensatz: die Charts

Die Datei `UK-top40-1964-1-2.tsv` enthält eine Ausgabe der britischen Top 40 aus dem Jahr 1964. Sie muss im selben Ordner liegen wie dieses Notebook.

Das Einlesen von Dateien ist Thema von Notebook 3. Die Zeile dafür ist deshalb vorgegeben.

In [138]:
# sep="\t": Die Spalten sind in dieser Datei durch Tabulatoren getrennt
charts = pd.read_csv("csv & excel Dateien/UK-top40-1964-1-2.tsv", sep="\t")
charts.head()

,Pos,LW,Title,Artist,Publisher,Peak Pos,WoC
0,1,1,I WANT TO HOLD YOUR HAND,THE BEATLES,PARLOPHONE,1,5
1,2,6,GLAD ALL OVER,THE DAVE CLARK FIVE,COLUMBIA,2,7
2,3,2,SHE LOVES YOU,THE BEATLES,PARLOPHONE,1,19
3,4,3,YOU WERE MADE FOR ME,FREDDIE AND THE DREAMERS,COLUMBIA,3,9
4,5,9,TWENTY FOUR HOURS FROM TULSA,GENE PITNEY,UNITED ARTISTS,5,5


### Übung 10 – Die Verlage der Charts (zu Block D und E, Häufigkeitstabellen)

- **a)** Erstelle für die Spalte `Publisher` eine Häufigkeitstabelle mit absoluter und relativer Häufigkeit.
- **b)** Welche Publisher teilen sich Platz 1, und welchen Anteil hat jeder von ihnen?
- **c)** Welches Skalenniveau hat `Publisher`?
- **d)** Warum wäre eine kumulierte Häufigkeit hier nicht sinnvoll?

Beantworte b) bis d) als Kommentar.

In [139]:
# a)


In [140]:
# b) bis d)


## 21 – Zweiter Schauplatz: die TechAcademy GmbH

Die TechAcademy GmbH ist ein IT-Ausbildungsunternehmen. Führe die folgende Zelle aus. Sie erstellt zwei DataFrames: `bewerbungen` und `kurse`.

In [141]:
# Bewerbungen für IT-Kurse der TechAcademy GmbH
bewerbungen = pd.DataFrame({
    "Name":           ["Finn Müller", "Lea Wagner", "Omar Hassan", "Jana Becker", "Timo Schulz"],
    "Alter":          [24, 31, 28, 35, 22],
    "Wohnort":        ["Berlin", "Hamburg", "Berlin", "München", "Hamburg"],
    "Kurs":           ["Python Grundlagen", "Data Analysis", "SQL für Einsteiger",
                       "Python Grundlagen", "Linux Basics"],
    "Bewerbungsnote": [2.3, 1.7, 3.0, 1.0, 2.7],
})

# Kursangebot der TechAcademy GmbH
kurse = pd.DataFrame({
    "Kursname":    ["Python Grundlagen", "SQL für Einsteiger", "Data Analysis",
                    "Linux Basics", "Power BI"],
    "Kategorie":   ["Programmierung", "Datenbanken", "Programmierung",
                    "Betriebssysteme", "Business Intelligence"],
    "Preis_EUR":   [799, 599, 999, 499, 899],
    "Anmeldungen": [142, 98, 113, 54, 88],
})

print(bewerbungen)
print()
print(kurse)

          Name  Alter  Wohnort                Kurs  Bewerbungsnote
0  Finn Müller     24   Berlin   Python Grundlagen             2.3
1   Lea Wagner     31  Hamburg       Data Analysis             1.7
2  Omar Hassan     28   Berlin  SQL für Einsteiger             3.0
3  Jana Becker     35  München   Python Grundlagen             1.0
4  Timo Schulz     22  Hamburg        Linux Basics             2.7

             Kursname              Kategorie  Preis_EUR  Anmeldungen
0   Python Grundlagen         Programmierung        799          142
1  SQL für Einsteiger            Datenbanken        599           98
2       Data Analysis         Programmierung        999          113
3        Linux Basics        Betriebssysteme        499           54
4            Power BI  Business Intelligence        899           88


### Übung 11 – Bewerbungsdaten erkunden (zu Block A)

Arbeite mit `bewerbungen`.

- **a)** Zeige die ersten 3 Zeilen.
- **b)** Zeige die letzten 2 Zeilen.
- **c)** Wie viele Zeilen und Spalten hat der DataFrame?
- **d)** Gib die Spaltennamen als Liste aus.
- **e)** Lass dir die technische Zusammenfassung mit den Datentypen ausgeben.
- **f)** Erstelle eine statistische Zusammenfassung der Zahlenspalten. *Hinweis: `describe()` kommt in Notebook 2 genauer dran.*

In [142]:
# a)


In [143]:
# b)


In [144]:
# c) und d)


In [145]:
# e)


In [146]:
# f)

### Übung 12 – Gezielt auswählen (zu Block B)

Arbeite mit `bewerbungen`.

- **a)** Wähle mit `.iloc` die gesamte dritte Zeile aus.
- **b)** Wähle mit `.iloc` den Wert in der ersten Zeile und der dritten Spalte aus. Überlege vorher, welchen Wert du erwartest.
- **c)** Wähle mit `.iloc` die Zeilen an den Positionen 1 bis einschließlich 3 aus.
- **d)** Wähle mit `.loc` für die Zeilen mit den Labels 0 und 4 nur die Spalten `Name` und `Bewerbungsnote` aus.
- **e)** Wähle mit `.loc` die komplette Zeile mit dem Label 2 aus.
- **f)** Wähle mit `.loc` die gesamte Spalte `Wohnort` aus. *Hinweis: Ein Doppelpunkt steht für alle Zeilen.*

In [147]:
# a)


In [148]:
# b)


In [149]:
# c)


In [150]:
# d)


In [151]:
# e)

In [152]:
# f)

### Übung 13 – Bewerbungen filtern (zu Block C)

Arbeite mit `bewerbungen`.

- **a)** Zeige alle Personen, die älter als 27 sind und aus Berlin kommen.
- **b)** Welcher Anteil der Bewerbungen gilt dem Kurs „Python Grundlagen"? Gib ihn in Prozent aus.

In [153]:
# a)


In [154]:
# b)


### Übung 14 – Kursdaten bearbeiten (zu Block C)

Arbeite mit `kurse`. Weise jedes Ergebnis wieder `kurse` zu, damit die Änderung erhalten bleibt.

- **a)** Setze die Spalte `Kursname` als Index.
- **b)** Füge eine Spalte `Umsatz_EUR` hinzu: `Preis_EUR` mal `Anmeldungen`.
- **c)** Benenne `Preis_EUR` in `Kursgebuehr` und `Anmeldungen` in `Teilnehmerzahl` um.
- **d)** Erstelle einen DataFrame `teure_kurse` mit allen Kursen, deren `Kursgebuehr` über 700 EUR liegt.
- **e)** Entferne die Spalte `Kategorie` aus `kurse`.

In [155]:
# a)


In [156]:
# b)


In [157]:
# c)


In [158]:
# d)


In [159]:
# e)


### Übung 15 – Umsatz nach Preisgruppe (zu Block C)

Arbeite mit dem Ergebnis aus Übung 14. Berechne den Gesamtumsatz (`Umsatz_EUR`) getrennt für Kurse mit einer `Kursgebuehr` über 700 EUR und für Kurse mit 700 EUR oder weniger.

In [160]:
# Übung 15


## 22 – Zusatz: Series und DataFrames selbst bauen

Bisher kam jede Series aus einer Spalte der Umfrage. Du kannst Series und DataFrames auch direkt anlegen. Die Beispiele stammen von der TechAcademy GmbH.

**Eine Series aus einer Liste.** Ohne weitere Angabe vergibt Pandas den Index 0, 1, 2, ...

In [161]:
stundenlohn = pd.Series([55, 60, 48, 72, 65])
print(stundenlohn)

0    55
1    60
2    48
3    72
4    65
dtype: int64


**Eine Series mit eigenem Index.** Mit `index=` bekommt jeder Wert ein eigenes Label. Über das Label holst du einen einzelnen Wert heraus.

In [162]:
stundenlohn_id = pd.Series(
    [55, 60, 48, 72, 65],
    index=["T001", "T002", "T003", "T004", "T005"]
)

print(stundenlohn_id)
print()
print("Stundenlohn von T003:", stundenlohn_id["T003"])

T001    55
T002    60
T003    48
T004    72
T005    65
dtype: int64

Stundenlohn von T003: 48


**Ein DataFrame aus einem Dictionary.** Die Umfrage entstand aus einer Liste von Zeilen. Häufiger siehst du die Schreibweise mit einem Dictionary: Jeder Schlüssel ist ein Spaltenname, die Liste dahinter enthält die Werte dieser Spalte. So sind auch `bewerbungen` und `kurse` in Abschnitt 21 entstanden.

In [163]:
trainer_daten = {
    "Name":            ["Sandra Meier", "Tobias Lange", "Yara Özcan", "Michael Brandt", "Nina Fuchs"],
    "Fachbereich":     ["Python & Data", "Datenbanken", "Linux & Netzwerk", "Python & Data", "Business Intelligence"],
    "Einstelldatum":   ["2019-03-01", "2021-06-15", "2020-09-01", "2022-01-10", "2023-04-01"],
    "Stundenlohn_EUR": [55, 60, 48, 72, 65],
}

trainer = pd.DataFrame(trainer_daten)
trainer

,Name,Fachbereich,Einstelldatum,Stundenlohn_EUR
0,Sandra Meier,Python & Data,2019-03-01,55
1,Tobias Lange,Datenbanken,2021-06-15,60
2,Yara Özcan,Linux & Netzwerk,2020-09-01,48
3,Michael Brandt,Python & Data,2022-01-10,72
4,Nina Fuchs,Business Intelligence,2023-04-01,65


**Der Index gleich beim Erstellen.** Auch ein DataFrame nimmt `index=` an. Danach sprichst du Zeilen mit `.loc` über ihr Label an.

In [164]:
trainer_ids = ["T001", "T002", "T003", "T004", "T005"]
trainer_mit_id = pd.DataFrame(trainer_daten, index=trainer_ids)

print(trainer_mit_id)
print()
print(trainer_mit_id.loc["T003"])

                Name            Fachbereich Einstelldatum  Stundenlohn_EUR
T001    Sandra Meier          Python & Data    2019-03-01               55
T002    Tobias Lange            Datenbanken    2021-06-15               60
T003      Yara Özcan       Linux & Netzwerk    2020-09-01               48
T004  Michael Brandt          Python & Data    2022-01-10               72
T005      Nina Fuchs  Business Intelligence    2023-04-01               65

Name                     Yara Özcan
Fachbereich        Linux & Netzwerk
Einstelldatum            2020-09-01
Stundenlohn_EUR                  48
Name: T003, dtype: object


**Neue Spalten aus einer Liste und aus einer Bedingung.** Eine Liste braucht genau so viele Einträge, wie der DataFrame Zeilen hat. Eine Bedingung ergibt eine Spalte mit `True` und `False`.

In [165]:
trainer["Kurse_pro_Monat"] = [8, 10, 6, 12, 9]
trainer["Senior"] = trainer["Stundenlohn_EUR"] >= 60
trainer

,Name,Fachbereich,Einstelldatum,Stundenlohn_EUR,Kurse_pro_Monat,Senior
0,Sandra Meier,Python & Data,2019-03-01,55,8,False
1,Tobias Lange,Datenbanken,2021-06-15,60,10,True
2,Yara Özcan,Linux & Netzwerk,2020-09-01,48,6,False
3,Michael Brandt,Python & Data,2022-01-10,72,12,True
4,Nina Fuchs,Business Intelligence,2023-04-01,65,9,True


**Eine Spalte mit `del` löschen.** `del` entfernt die Spalte sofort und ohne Zuweisung. `drop(columns=...)` aus Abschnitt 10 ist flexibler, weil es mehrere Spalten auf einmal entfernen kann.

In [166]:
del trainer["Kurse_pro_Monat"]
print(trainer.columns.tolist())

['Name', 'Fachbereich', 'Einstelldatum', 'Stundenlohn_EUR', 'Senior']


### Übung 16 – Kursteilnehmer anlegen (zu Abschnitt 22)

Führe zuerst die Zelle mit dem Dictionary aus.

- **a)** Erstelle aus `teilnehmer_daten` einen DataFrame `teilnehmer`.
- **b)** Füge die Spalte `Kursgebuehr_EUR` mit den Werten 799, 599, 799, 999 und 499 hinzu.
- **c)** Füge eine Spalte `Note_Prozent` hinzu. Die Note 1,0 soll 100 % ergeben, die Note 5,0 soll 0 % ergeben: `(5 - Note) / 4 * 100`. Runde auf eine Nachkommastelle.
- **d)** Füge eine Spalte `Sehr_gut` hinzu. Sie soll `True` enthalten, wenn die Abschlussnote besser als 2,0 ist.
- **e)** Erstelle eine Series `kursnamen` mit den Kursen Python Grundlagen, SQL für Einsteiger, Data Analysis, Linux Basics und Power BI und dem Index K01 bis K05. Gib den Kurs mit dem Label K03 aus.

In [167]:
teilnehmer_daten = {
    "Name":           ["Mia Schulz", "Jonas Weber", "Lena Koch", "Tim Bauer", "Sara Yilmaz"],
    "Kurs":           ["Python Grundlagen", "SQL für Einsteiger", "Python Grundlagen", "Data Analysis", "Linux Basics"],
    "Eintrittsdatum": ["2024-02-01", "2024-02-01", "2024-03-15", "2024-03-15", "2024-04-01"],
    "Abschlussnote":  [2.0, 1.7, 3.3, 1.0, 2.7],
    "Bestanden":      [True, True, True, True, True],
}

In [168]:
# a)

In [169]:
# b)

In [170]:
# c)

In [171]:
# d)

In [172]:
# e)

In [173]:
# f)

---
## 23 – Zusammenfassung

| Aufgabe | Befehl |
|---|---|
| Überblick | `head()`, `tail()`, `shape`, `columns`, `info()` |
| Spalte auswählen | `df["spalte"]`, `df[["a", "b"]]` |
| Kennzahlen | `min()`, `max()`, `sum()`, `count()`, `mean()` |
| Zeilen nach Position oder Label | `df.iloc[...]`, `df.loc[...]` |
| Filtern | `df[bedingung]`, `&`, `\|`, `isin([...])` |
| Anteil | `(bedingung).mean()` |
| Zufallsstichprobe | `sample(n=..., random_state=...)` |
| Spalten bearbeiten | Zuweisung, `rename(columns=...)`, `drop(columns=...)` |
| Index | `set_index()`, `reset_index()` |
| Datentypen und Ausprägungen | `dtypes`, `unique()`, `nunique()` |
| Geordnete Kategorie | `pd.Categorical(..., categories=..., ordered=True)` |
| Sortieren | `sort_values()`, `sort_index()` |
| Häufigkeiten | `value_counts()`, `value_counts(normalize=True)`, `cumsum()` |
| Klassen | `pd.cut(..., bins=...)` |
| Kreuztabelle | `pd.crosstab(a, b, margins=True, normalize=...)` |
| Series und DataFrame anlegen | `pd.Series([...], index=[...])`, `pd.DataFrame({...}, index=[...])` |
| Spalte sofort löschen | `del df["spalte"]` |

---
*Material erstellt von Julia Brutskaya*